### PR TITLE
Fix update treasury with delegate rewards bug

### PR DESCRIPTION
**Pull request by prat-gpt**
# Description of change
## High level
* Update treasury with unused delegation rewards when there are no funding votes.

# Description of bug or vulnerability and solution
* `UpdateTreasury` method always updated the treasury with delegate rewards whether or not the user has voted in the funding stage which led to the possibility of treasury amount more than the token balance>
* Fixed this by updating treasury with delegate rewards only when no users vote in the funding stage.
* https://github.com/code-423n4/2023-05-ajna-findings/issues/461
* https://github.com/code-423n4/2023-05-ajna-findings/issues/450
* https://github.com/code-423n4/2023-05-ajna-findings/issues/289
* https://github.com/code-423n4/2023-05-ajna-findings/issues/263

### CODE REVIEW

**Review by MikeHathaway — COMMENTED** [review]
LGTM - can you check if the multipleDistribution and Finalize invariants need to be updated?

File: test/unit/StandardFunding.t.sol [hunk 1]
```diff
@@ -1157,6 +1157,184 @@ contract StandardFundingGrantFundTest is GrantFundTestHelper {
         _grantFund.claimDelegateReward(distributionId);
     }
 
+    function testTreasureUpdatedWithSurplusTokens() external {
```
Comment (MikeHathaway) [inline test/unit/StandardFunding.t.sol]: Suggest renaming "Treasure" -> "Treasury"

**Review by MikeHathaway — APPROVED** [review]
LGTM

File: src/grants/base/StandardFunding.sol [hunk 2]
```diff
@@ -203,18 +203,22 @@ abstract contract StandardFunding is Funding, IStandardFunding {
...
             unchecked { ++i; }
         }
 
-        // re-add non distributed tokens to the treasury
-        treasury += (fundsAvailable - totalTokensRequested);
+        uint256 totalDelegateRewards;
+        // Increament totalTokenDistributed by delegate rewards if anyone has voted during funding voting
+        if (_distributions[distributionId_].fundingVotePowerCast != 0) totalDelegateRewards = (fundsAvailable / 10);
+
+        // readd non distributed tokens to the treasury
```
Comment (ith-harvey) [inline src/grants/base/StandardFunding.sol]: NIT: `re-add`

File: src/grants/base/StandardFunding.sol [hunk 3]
```diff
@@ -203,18 +203,22 @@ abstract contract StandardFunding is Funding, IStandardFunding {
...
-            totalTokensRequested += _standardFundingProposals[fundingProposalIds[i]].tokensRequested;
+            totalTokenDistributed += _standardFundingProposals[fundingProposalIds[i]].tokensRequested;
 
             unchecked { ++i; }
         }
 
-        // re-add non distributed tokens to the treasury
-        treasury += (fundsAvailable - totalTokensRequested);
+        uint256 totalDelegateRewards;
+        // Increament totalTokenDistributed by delegate rewards if anyone has voted during funding voting
```
Comment (ith-harvey) [inline src/grants/base/StandardFunding.sol]: NIT: spelling error Increment

File: test/invariants/INVARIANTS.md [hunk 4]
```diff
@@ -9,7 +9,7 @@
 - #### Distribution Period:
     - **DP1**: Only one distribution period should be active at a time. Each successive distribution period's start block should be greater than the previous periods end block.
     - **DP2**: Each winning proposal successfully claims no more that what was finalized in the challenge stage
-    - **DP3**: A distribution's fundsAvailable should be equal to 2% of the treasurie's balance at the block `startNewDistributionPeriod()` is called.
+    - **DP3**: A distribution's fundsAvailable should be equal to 3% of the treasurie's balance at the block `startNewDistributionPeriod()` is called.
```
Comment (ith-harvey) [inline test/invariants/INVARIANTS.md]: NIT: `treasury's`

**Review by ith-harvey — COMMENTED** [review]
This PR successfully resolves the two outstanding issues caught by `461` & `289`

Noticing however that `totalDelegateRewards` does not account for the delegates who have not voted in the delegation period. If a single delegate votes in both the funding and screening rounds however all other delegates only vote in funding (should not receive a delegate reward) it appears we set aside their rewards even though they can never be claimed.

This permanently locks those unclaimable rewards in the contract, I will confirm in standup but I think this is expected functionality.

**Review by ith-harvey — APPROVED** [review]
LGTM once NIT fixes are made.

**Pull request merged**